compiler/rustc_data_structures/src/sync.rs

```diff
@@ -364,6 +338,49 @@ cfg_if! {
     }
 }
 
+#[derive(Debug)]
+pub struct WorkerLocal<T> {
+    single_thread: bool,
+    inner: T,
+    mt_inner: Option<rayon_core::WorkerLocal<T>>,
+}
+
+impl<T> WorkerLocal<T> {
+    /// Creates a new worker local where the `initial` closure computes the
+    /// value this worker local should take for each thread in the thread pool.
+    #[inline]
+    pub fn new<F: FnMut(usize) -> T>(mut f: F) -> WorkerLocal<T> {
+        if !PARALLEL.load(Ordering::Relaxed) {
+            WorkerLocal { single_thread: true, inner: f(0), mt_inner: None }
+        } else {
+            // Safety: `inner` would never be accessed when multiple threads
+            WorkerLocal {
+                single_thread: false,
+                inner: unsafe { MaybeUninit::uninit().assume_init() },
```

Comment:
This seems dangerous...
I am not entirely sure if the semantics of this have been fixed yet, but reading https://github.com/rust-lang/unsafe-code-guidelines/blob/master/active_discussion/validity.md it seems like this would be UB under that (since the assignment does a typed copy at type `T`, which does not allow uninit).
Either way, this is certainly a pattern that is discouraged, and it seems like the compiler should set an example here...
It seems much better to use a `union` between `inner` and `mt_inner` here, since it is guaranteed to only access the right field. (Or even better, an enum, since `single_thread` then functions as a discriminant... which basically makes it a homegrown enum anyways)

Reply:
Yeah this is very unsound because rustc emits `noundef` for almost all types. So this is immediately UB from LLVM's PoV, and the current thought for Rust rules with out any real thoughts on it not.

Reply:
It makes sense. I am a little worried about the efficiency of using enum or union, maybe it is better to use inner: Option<T> here

Reply:
I don't quite understand what the efficiency problem of an enum or union would be - checking the discriminant of an enum should be basically the same as checking `if self.single_thread {`, right?
(Additionally, using an `Option` would add increase the size by adding the `Option`s discriminant in addition to `single_thread`)
I have opened #109528 to test the performance of my suggestion.

Reply:
But `Option` is also an enum, so it should have the same effect?

Reply:
We can use `unwrap` which is a const function when signle_thread so i guess it is relatively more efficient

Reply:
I'd suggest instead of `T` to use `MaybeUninit<T>` as the type during storage. This avoids the segfault on `Drop`. Of course you'd have to impl `Drop` manually then.